rat/codegen_backends.py

```diff
@@ -1,357 +1,269 @@
-                        carry_size = max(carry_size, *[shift.value for shift in shifts])
-
-                self.expression_string += f"# Assigned param: {self.lhs_key}\n"
-                if self.lhs_used_in_rhs:
-                    lhs_record = self.variable_table[self.lhs_key]
-                    lhs_size = lhs_record.base_df.shape[0]
-
-                    zero_tuple_string = f"({','.join(['0.0'] * carry_size)})"
-                    carry_strings = [f"carry{n}" for n in range(1, carry_size + 1)]
-                    carry_tuple_string = f"({','.join(carry_strings)})"  # (carry0, carry1, ...)
-                    next_carry_tuple_string = f"({','.join(['next_value'] + carry_strings[:-1])})"
-                    scan_function_name = f"scan_function_{self.variable_table.get_unique_number()}"
-
-                    self.expression_string += f"def {scan_function_name}(carry, index):\n"
-                    if need_first_in_group_indicator:
-                        self.expression_string += f"    carry = jax.lax.cond(first_in_group_indicators['{self.lhs_key}'][index], lambda : {zero_tuple_string}, lambda : carry)\n"
-                    self.expression_string += f"    {carry_tuple_string} = carry\n"
-                    self.expression_string += f"    next_value = "
-                    self.at_rhs_of_scan = True
-                    self.generate(rhs)
-                    self.at_rhs_of_scan = False
-                    self.expression_string += "\n"
-                    self.expression_string += f"    return {next_carry_tuple_string}, next_value\n\n"
-
-                    self.expression_string += "_, "
-                    self.generate(lhs)
-                    self.expression_string += f" = jax.lax.scan({scan_function_name}, {zero_tuple_string}, jax.numpy.arange({lhs_size}))"
-                    self.expression_string += "\n"
+                    index_lookup = ""
+
+                return f"parameters['{node.name}']{index_lookup}"
 
+    def walk_Literal(self, node: ast.Literal):
+        return f"{node.value}"
+
+
+class CodeExecutor(NodeWalker):
+    arguments: Dict[ast.Variable, Any]
+    parameters: Dict[str, Any]
+    left_side_of_sampling: ast.ModelBase
+
+    def __init__(self, arguments: Dict[ast.Variable, Any] = None, parameters: Dict[str, Any] = None):
+        self.arguments = arguments
+        self.parameters = parameters
+        self.left_side_of_sampling = None
+
+    def walk_Statement(self, node: ast.Statement):
+        if node.op == "~":
+            self.left_side_of_sampling = node.left
+            return_value = self.walk(node.right)
+            self.left_side_of_sampling = None
+            return return_value
+        else:
+            raise CompileError(f"{node.op} operator not supported in BaseCodeGenerator", node)
+
+    def walk_Logical(self, node: ast.Logical):
+        left = self.walk(node.left)
+        right = self.walk(node.right)
+        return eval(f"(left {node.op} right)")
```

Comment:
Eval acts just like running code I think. It can see that left and right have been defined lol. It's a bit funky but works.

I use `eval(node.name)` for functions. This basically uses eval to do a string lookup by function-name for whatever it should be executing. Someone on Stackoverflow showed this.

We could probably do something less open-to-mistakes than this but it was easy to code up :P.